### PR TITLE
#[UIC-1170]|Design PR : Intra dashboard interaction support

### PR DESCRIPTION
**Pull request by jitendra-kumawat**
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Add support for Intra Dashboard Interactivity support in superset. 
Intra Dashboard Interactivity means whenever user do selection over any visualization/chart in Dashboard then selection of that chart can apply to rest/set of charts in following manner depends on ux/usecases-
1. apply  as  filters/global filters in sql query as where clause with required query operators referred as **applyFilter**
2. change  model or schema of  chart referred as **changeConfig**
3. show information in popup  referred as **showPopup**
4. redirect to new URL or any other dashboard url referred as **redirectToUrl**
5. mix of any action listed above referred as **applyFilter and configChange**
6. selection sync or highlight same selections in different charts referred as **highlightSelections**

so to solve above use cases here  I am proposing a Design PR to implement these feature in superset and calling it as Publish Subscriber Design Pattern , in this any charts which is used for selection (via click or user-interaction) called as Publisher and rest /set of charts listening these selections and update itself based on action defined  called as subscribers.
refer to following terminology 
Publisher - any chart publish selection on click
Publish Columns - list of columns and respective data published in form of selection 

Subscriber- any chart subscribe selections published by publisher
Action - action need to execute on subscription, actions can be any above use case listed in introduction (1-6)
Operator- SQL query operators used while creating where clause  like =,!=,IN etc
Subscribe Columns- subscriber can choose columns from publish columns before execute action
Extra- store extra metadata required  like publish value can map with chart ids to dynamic chart changes

Interactor - acts mediator between charts and dashboard to  identify publisher, subscriber and
execute action with correct information

PubSubMetaData - metadata info related to publisher and subscriber 

**Over all  design  flow diagram**
Screenshot 2019-05-13 at 12 28 36 PM

**UI for get info of publishers and subscribers from user and store as PubSubMetaData**

Screenshot 2019-05-13 at 12 32 00 PM

as per these diagram we need to introduce separate  UI  at dashboard level to get PubSub info from USER and store as metadata but we have already implemented basic publish and subscriber design in superset to solve  applyFilter  interaction at dashboard level ( https://github.com/Guavus/incubator-superset/pull/19 ), so we will implement above design in phases to resolve migration issues-

Phase -I

1. All Pub sub info will be store at Chart level in form data and new options will be defined and added in PubSub section of Chart
2. PubSubMetaData will be created from slices and store at dashboard level and Options of **Reconcile** will be provided at  dashboard level
3. once user will choose this option PubSubMetaData will create and store
4. interactor will be implemented as per design.  

Screenshot 2019-05-13 at 12 57 14 PM

**Low Level Data Structure and Properties
IMG_20190513_164645
IMG_20190513_164936

**

Phase -2
1. Designed  UI  will be implemented as per design and read all pubsub info from UI instead slices formdata
2. **Reconcile** option change to **Configure Filters**   and update functionality

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
NA

### TEST PLAN
NA

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [x] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@bipinsoniguavus 
@arpit-agarwal 
@ankursinghal2005 
@rasmi-ranjan-guavus

### CODE REVIEW

**Comment by arpit-agarwal** [comment]
LGTM. However can we close low level data-structure contract with relevant implementors of various sub-tasks

**Comment by ankursinghal2005** [comment]
LGTM. We have discussed low level meta data structure internally. We can go ahead..

File: superset/assets/src/dashboard/actions/dashboardState.js [hunk 1]
```diff
@@ -135,6 +135,23 @@ export function onSave() {
...
+
+export const ON_FINISH_RECONCILE = 'ON_FINISH_RECONCILE';
+export function onFinishReconcile(publishSubscriberMap) {
+  return { type: ON_FINISH_RECONCILE ,publishSubscriberMap};
+}
+
+export function reconcileSuccess(data,id){
+  return dispatch => {
+    dispatch(onFinishReconcile(data.pub_sub_info))
+    dispatch(saveDashboardRequest(data,id,SAVE_TYPE_OVERWRITE,'This dashboard was reconcile successfully.'));
```
Comment (ankursinghal2005) [inline superset/assets/src/dashboard/actions/dashboardState.js]: keep space after ever common to remove lint errors.

File: superset/assets/src/dashboard/actions/dashboardState.js [hunk 2]
```diff
@@ -135,6 +135,23 @@ export function onSave() {
...
 }
 
+export const ON_RECONCILE = 'ON_RECONCILE';
+export function onReconcile() {
+  return { type: ON_RECONCILE };
+}
+
+export const ON_FINISH_RECONCILE = 'ON_FINISH_RECONCILE';
+export function onFinishReconcile(publishSubscriberMap) {
+  return { type: ON_FINISH_RECONCILE ,publishSubscriberMap};
```
Comment (ankursinghal2005) [inline superset/assets/src/dashboard/actions/dashboardState.js]: Add space after ever comma

File: superset/assets/src/dashboard/reducers/dashboardState.js [hunk 3]
```diff
@@ -30,10 +30,18 @@ import {
...
+  ON_RECONCILE,
+  ON_FINISH_RECONCILE,
 } from '../actions/dashboardState';
 
 export default function dashboardStateReducer(state = {}, action) {
   const actionHandlers = {
+    [ON_RECONCILE]() {
+      return { ...state, doReconcile: true };
+    },
+    [ON_FINISH_RECONCILE]() {
```
Comment (ankursinghal2005) [inline superset/assets/src/dashboard/reducers/dashboardState.js]: we can use replace word **ON_FINISH_RECONCILE** Finish word with success. Success key word looks good.

File: superset/assets/src/dashboard/util/getPublishSubscriberMap.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+                subs = {};
+            }
+            subs[element.id] = createSubscriberData(element, publishers)
+        }
+    });
+    return subs;
+}
+
+export default function getPublishSubscriberMap(charts) {
+    var publishers = getPublishers(charts);
```
Comment (ankursinghal2005) [inline superset/assets/src/dashboard/util/getPublishSubscriberMap.js]: could we rename getPublishers to getPublishersMap && getSubscribers to getSubscribersMap. Actually we are getting mapping of each.

File: superset/assets/src/dashboard/util/getPublishSubscriberMap.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+}
+
+function isSubscriber(chart) {
+    return (chart.formData.linked_slice && chart.formData.linked_slice.length > 0)
+}
+
+function getSubscribers(charts, publishers) {
+    var subs = undefined;
```
Comment (ankursinghal2005) [inline superset/assets/src/dashboard/util/getPublishSubscriberMap.js]: no need to assigned value as undefined. By default variable value is undefined.

File: superset/assets/src/dashboard/util/getPublishSubscriberMap.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    charts.forEach(element => {
+        if (isSubscriber(element)) {
+            if (subs == undefined) {
```
Comment (ankursinghal2005) [inline superset/assets/src/dashboard/util/getPublishSubscriberMap.js]: we can use if(!sub) condition except checking value is undefined

File: superset/assets/src/dashboard/util/getPublishSubscriberMap.js [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+function isPublisherExistInLinkedSlices(slice, publisherId) {
+    var item = _.find(slice.formData.linked_slice ,function(item) {
+        return  ((Number.isInteger(item) && item == publisherId ) ||
+        (item instanceof Object && item.publisher_id == publisherId))
+    })
+    return (item != undefined)
+}
+
+function getLinkedSlices(slices, publishers) {
```
Comment (ankursinghal2005) [inline superset/assets/src/dashboard/util/getPublishSubscriberMap.js]: we are passing linked_slices in param. So please rename slices to linkedSlices.

File: superset/assets/src/dashboard/util/getPublishSubscriberMap.js [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        }
+        // as per new ds
+        else if (element instanceof Object && publishers.hasOwnProperty(element.publisher_id)) {
+            ls[element.publisher_id] = element.subscribe_columns;
+        }
+    });
+    return ls;
+}
+
+function createSubscriberData(chart, publishers) {
```
Comment (ankursinghal2005) [inline superset/assets/src/dashboard/util/getPublishSubscriberMap.js]: we should rename **createSubscriberData** to createSubscriberSliceData. We are creating every slice subscriber data.

File: superset/assets/src/dashboard/util/getPublishSubscriberMap.js [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import * as _ from 'lodash';
+const APPLY_FILTER = 'APPLY_FILTER';
+
+function createPublishData(chart, charts) {
+    return {
```
Comment (ankursinghal2005) [inline superset/assets/src/dashboard/util/getPublishSubscriberMap.js]: we should rename createPublishData to createPublishSliceData. We are creating every slice publish data.

File: superset/assets/src/dashboard/util/getPublishSubscriberMap.js [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        id: chart.id,
+        publish_columns: chart.formData.publish_columns,
+        subcribers: getSubsribersFor(chart.id, charts),
+        viz_type: chart.formData.viz_type
+    }
+
+}
+
+function getSubsribersFor(publisherId, charts) {
```
Comment (ankursinghal2005) [inline superset/assets/src/dashboard/util/getPublishSubscriberMap.js]: we are retuning subscriber slices id's. It is not much clear from function name.

File: superset/assets/src/dashboard/components/Dashboard.jsx [hunk 11]
```diff
@@ -171,7 +173,22 @@ class Dashboard extends React.PureComponent {
   }
 
   componentDidUpdate(prevProps) {
-    const { refresh, filters, hasUnsavedChanges } = this.props.dashboardState;
+    const { refresh, filters, hasUnsavedChanges, doReconcile } = this.props.dashboardState;
```
[the remark this answers is not in the document]
Comment (jitendra-kumawat) [reply superset/assets/src/dashboard/components/Dashboard.jsx]: done

File: superset/assets/spec/javascripts/dashboard/components/HeaderActionsDropdown_spec.jsx [hunk 12]
```diff
@@ -66,9 +67,9 @@ describe('HeaderActionsDropdown', () => {
       expect(wrapper.find(SaveModal)).toHaveLength(0);
     });
 
-    it('should render one MenuItem', () => {
+    it('should render 2 MenuItem', () => {
```
[the remark this answers is not in the document]
Comment (jitendra-kumawat) [reply superset/assets/spec/javascripts/dashboard/components/HeaderActionsDropdown_spec.jsx]: done

File: superset/assets/src/dashboard/actions/dashboardState.js (hunk 1 above)
Comment (jitendra-kumawat) [reply superset/assets/src/dashboard/actions/dashboardState.js]: done

File: superset/assets/src/dashboard/actions/dashboardState.js (hunk 2 above)
Comment (jitendra-kumawat) [reply superset/assets/src/dashboard/actions/dashboardState.js]: done

File: superset/assets/src/dashboard/reducers/dashboardState.js (hunk 3 above)
Comment (jitendra-kumawat) [reply superset/assets/src/dashboard/reducers/dashboardState.js]: renamed to ON_SUCCESS_RECONCILE

File: superset/assets/src/dashboard/util/getPublishSubscriberMap.js (hunk 9 above)
Comment (jitendra-kumawat) [reply superset/assets/src/dashboard/util/getPublishSubscriberMap.js]: renamed to createPublishDataFor

File: superset/assets/src/dashboard/util/getPublishSubscriberMap.js (hunk 8 above)
Comment (jitendra-kumawat) [reply superset/assets/src/dashboard/util/getPublishSubscriberMap.js]: done

File: superset/assets/src/dashboard/util/getPublishSubscriberMap.js (hunk 7 above)
Comment (jitendra-kumawat) [reply superset/assets/src/dashboard/util/getPublishSubscriberMap.js]: done

File: superset/assets/src/dashboard/util/getPublishSubscriberMap.js (hunk 6 above)
Comment (jitendra-kumawat) [reply superset/assets/src/dashboard/util/getPublishSubscriberMap.js]: done

File: superset/assets/src/dashboard/util/getPublishSubscriberMap.js (hunk 5 above)
Comment (jitendra-kumawat) [reply superset/assets/src/dashboard/util/getPublishSubscriberMap.js]: done

File: superset/assets/src/dashboard/util/getPublishSubscriberMap.js (hunk 4 above)
Comment (jitendra-kumawat) [reply superset/assets/src/dashboard/util/getPublishSubscriberMap.js]: done

**Pull request merged**